src/sender_queue/mod.rs

```diff
@@ -41,9 +41,9 @@ pub trait SenderQueueableOutput<N, M>
 where
     N: NodeIdT,
 {
-    /// Returns an optional new node added with the batch. This node should be added to the set of
-    /// all nodes.
-    fn added_node(&self) -> Option<N>;
+    /// Returns the new set of validator that this batch is starting key generation for, including
+    /// the existing ones. These should be added to the set of all nodes.
```

Comment:
Good point… that's a more general issue that I'm still unclear about:
Do we want to have the sender queue automatically manage the set of observer nodes or should the user also have some control over that (to e.g. add an observer that isn't about to join as a validator)?
* In the first case, we'd probably also want to _remove_ peers on `ChangeState::Complete`, in addition to adding them on `ChangeState::InProgress`, so that the set of peers is always exactly the union of the current and the next set of validators.
* In the second case, I'm not sure: Either we'd remove this feature completely and _require_ the user to add and remove observers, or we'd maintain an additional list of observers in `SenderQueue` (those would be the ones that DHB will never know about; they only observe, without ever becoming a validator).

For now, maybe I should actually implement the first point right away? It probably wouldn't be much work and I think it's mainly the intended behavior of the current code?

Reply:
Our sender queue is a reference implementation which, I think, should be as simple as possible for that reason. I tend to choose the second option. In principle, the user can fully control the peers and possibly whether or not new peers join as observers only. Removing old peers however should be totally fine in our implementation. The user can change that in their implementation.

Reply:
I just realized that removing peers is actually more complicated than I thought:
If we remove a peer in epoch 42, we'll still need to deliver their messages from epochs up to 41. I think that'll require some changes to `update_epoch`? I'm not entirely sure. Maybe we should do that as a separate PR after all?

Reply:
Yes, I think that would work. I created https://github.com/poanetwork/hbbft/issues/341 for it.